lib/ansible/modules/network/fortimanager/fmgr_fwpol_package.py

```diff
@@ -0,0 +1,543 @@
+        "scope": members_list
+    }
+    # EXECUTE THE INSTALL REQUEST
+    url = '/securityconsole/install/package'
+    response = fmg.execute(url, datagram)
+    return response
+
+
+def main():
+    argument_spec = dict(
+        adom=dict(required=False, type="str", default="root"),
+        host=dict(required=True, type="str"),
+        username=dict(fallback=(env_fallback, ["ANSIBLE_NET_USERNAME"])),
+        password=dict(fallback=(env_fallback, ["ANSIBLE_NET_PASSWORD"]), no_log=True),
+        mode=dict(choices=["add", "set", "delete"], type="str", default="add"),
+
+        name=dict(required=False, type="str"),
+        type=dict(required=False, type="str", choices=['pkg', 'folder', 'install']),
+        package_folder=dict(required=False, type="str"),
+        central_nat=dict(required=False, type="str", default="disable", choices=['enable', 'disable']),
+        fwpolicy_implicit_log=dict(required=False, type="str", default="disable", choices=['enable', 'disable']),
+        fwpolicy6_implicit_log=dict(required=False, type="str", default="disable", choices=['enable', 'disable']),
+        inspection_mode=dict(required=False, type="str", default="flow", choices=['flow', 'proxy']),
+        ngfw_mode=dict(required=False, type="str", default="profile-based", choices=['profile-based', 'policy-based']),
+        ssl_ssh_profile=dict(required=False, type="str"),
+        scope_members=dict(required=False, type="str"),
+        scope_members_vdom=dict(required=False, type="str", default="root"),
+        parent_folder=dict(required=False, type="str"),
+
+    )
+
+    module = AnsibleModule(argument_spec, supports_check_mode=True, )
+
+    # MODULE DATAGRAM
+    paramgram = {
+        "adom": module.params["adom"],
+        "name": module.params["name"],
+        "mode": module.params["mode"],
+        "type": module.params["type"],
+        "package-folder": module.params["package_folder"],
+        "central-nat": module.params["central_nat"],
+        "fwpolicy-implicit-log": module.params["fwpolicy_implicit_log"],
+        "fwpolicy6-implicit-log": module.params["fwpolicy6_implicit_log"],
+        "inspection-mode": module.params["inspection_mode"],
+        "ngfw-mode": module.params["ngfw_mode"],
+        "ssl-ssh-profile": module.params["ssl_ssh_profile"],
+        "scope_members": module.params["scope_members"],
+        "scope_members_vdom": module.params["scope_members_vdom"],
+        "parent_folder": module.params["parent_folder"],
+    }
+
+    # VALIDATE REQUIRED ARGUMENTS ARE PASSED; NOT USED IN ARGUMENT_SPEC TO ALLOW PARAMS TO BE CALLED FROM PROVIDER
+    # CHECK IF PARAMS ARE SET
+    if module.params["host"] is None or module.params["username"] is None:
+        module.fail_json(msg="Host and username are required for connection")
+
+    # CHECK IF LOGIN FAILED
+    fmg = AnsibleFortiManager(module, module.params["host"], module.params["username"], module.params["password"])
+    response = fmg.login()
```

Comment:
Looks like API login and http request/response handling is done outside of ansible and done through library 'pyFMG'. Essentially this means all fortimanager module need to provide username/password in all tasks individually   ! This does not look a clean design IMO. I would request to take a look at httapi plugin which will make modules will look more simple. In that case you would be able to provide username/password only in inventory (which also enables to use vault to store passwords in encrypted way ). You can also enable all connection settings like timeout etc to manage HTTP timeouts. we did similar things with another firewall module . Please have a look at lib/ansible/plugins/httpapi/ftd.py

Reply:
We value your feedback, thank you! We're looking into this for future modules. For now we'd like to move ahead with this architecture for these PR's we have in place. 

We've made moves on this already, so expect a solution to be implemented soon on future PRs.

Reply:
Hi @ftntcorecse 
Unfortunately we cannot accept the modules as-is. Until the modules are modified to clean design which is
a) Move the connection handling to connection plugin.
b) Modularity- Common functionality which has been duplicated in so many modules should be moved to module_utils/network/fortimanager and should not be repeated
c) Idempotency should be handled
d) Probably we don't need one module for each FortiManager operations unless we can not handle some corner cases of idempotency. We should have only bunch of modules which will do REST operations on ANY object.
Specific use-case should go to galaxy role.

Until the modules are modified to best practices listed, they can be posted to Ansible Galaxy in the meantime.
Thanks,
Deepak

Reply:
Hi @gdpak 
We've submitted PR # 49391 to address issue B) above. 

Can you please elaborate on C) ? Idempotency is handled by FortiManager itself. You can run our playbooks over and over and over and it will have the same effect. I can understand your concern, seeing how you probably don't have an instance of FortiManager to try this on. I'd be happy to stand up an instance of FortiManager you can access, so you can test and see yourself, or if you'd prefer to have a Go To Meeting we can meet and discuss further. 

RE: D) A year ago we started to develop for Ansible and we brought this up. We were told to logically group functions into modules and not blindly do edge cases, but also told not to make the modules so massive. This is a very subjective thing unfortunately, and we DO actually take time to group modules into logical functions, so maybe we can have a quick call to help establish some guidelines around this concern?

We're working on Issue A) moving the connection to the plugins. Look for something next week.

Reply:
@ftntcorecse It would be better to get together on a call and make a mutually agreed upon design before refactoring. please give me your email-id or you can send me an email about timings/invite at dagrawal@redhat.com.  Thanks